### PR TITLE
Retry deployments

### PR DESCRIPTION
**Pull request by RamIdeas**
## What this PR solves / how to test

This PR adds retries within `wrangler deploy` and `wrangler versions upload` to workaround spotty network connections and service flakes

We need to be careful about retrying the whole multi-step process because of some non-atomic steps.

Users have reported flakes at the upload step so we can safely add retries for that step to address those specific failures.

Fixes #0000

## Author has addressed the following

- Tests
  - [ ] TODO (before merge)
  - [x] Tests included
  - [ ] Tests not necessary because:
- E2E Tests CI Job required? (Use "e2e" label or ask maintainer to run separately)
  - [ ] I don't know
  - [ ] Required
  - [x] Not required because: covered by other tests
- Changeset ([Changeset guidelines](https://github.com/cloudflare/workers-sdk/blob/main/CONTRIBUTING.md#changesets))
  - [ ] TODO (before merge)
  - [x] Changeset included
  - [ ] Changeset not necessary because:
- Public documentation
  - [ ] TODO (before merge)
  - [ ] Cloudflare docs PR(s): 
  - [x] Documentation not necessary because: transparent change

### CODE REVIEW

**Comment by jahands** [comment]
Thank you for this!! The flakiness in deploying made my deploy CI job so unreliable (to the point that I added a retry wrapper script around `wrangler deploy`.) This will help a lot <3

**Review by andyjessop — APPROVED** [review]
Very nice 👍 

One question I have is do we know this is going to fix the problem, Or will it just make failed runs take longer?

File: packages/wrangler/src/__tests__/versions/versions.upload.test.ts [hunk 1]
```diff
@@ -37,11 +37,16 @@ describe("versions upload", () => {
...
 	}
-	function mockUploadVersion(has_preview: boolean) {
+	function mockUploadVersion(has_preview: boolean, flakeCount = 1) {
 		msw.use(
 			http.post(
 				`*/accounts/:accountId/workers/scripts/:scriptName/versions`,
 				({ params }) => {
+					if (flakeCount > 0) {
+						flakeCount--;
+						// return HttpResponse.error();
```
[the remark this answers is not in the document]
Comment (RamIdeas) [reply packages/wrangler/src/__tests__/versions/versions.upload.test.ts]: argh how? what??

Don't know how that got left uncommented. Fixed now.

**Comment by jahands** [comment]
> Very nice 👍
> 
> One question I have is do we know this is going to fix the problem, Or will it just make failed runs take longer?

I added up to 3 retries to my deployments months ago and have never had a flake since. Before that, it seemed like it had a 30% chance of failure due to 5XX errors.

**Comment by RamIdeas** [comment]
> Very nice 👍
> 
> One question I have is do we know this is going to fix the problem, Or will it just make failed runs take longer?

Failures due to flakey internet or backend will be fixed by this. This will improve the happy path for successful deployments. If there is a legit reason for failure, we can iterate on this solution to bail out of the retries – but I think for now we should optimise for the happy path here and take the hit of a few seconds on the sad path.

File: packages/wrangler/src/utils/retry.ts [hunk 2]
```diff
@@ -0,0 +1,18 @@
+import { setTimeout } from "node:timers/promises";
+
+export async function retryOnError<T>(
+	action: () => T | Promise<T>,
+	backoff = 2_000,
```
Comment (threepointone) [inline packages/wrangler/src/utils/retry.ts:5]: this signature means you have to pass a `backoff` if you want to define `attempts`, which isn't ideal. Can we put backoff and attempts into an `options` object instead, with both fields being optional?

File: packages/wrangler/src/utils/retry.ts [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+	attempts = 3
+): Promise<T> {
+	try {
+		return await action();
+	} catch (err) {
+		if (attempts <= 1) {
+			throw err;
+		}
+
+		await setTimeout(backoff);
```
Comment (threepointone) [inline packages/wrangler/src/utils/retry.ts:15]: should the default be 0? what would a 2 second wait (by default) accomplish?

**Pull request merged**